### PR TITLE
ci: add coverity scan

### LINKED ISSUE #46: Add a static analyzer to CI
By running a static analyzer tool during CI we could find dead code, faulty initiations and more.
Options might be Coverity or Clang static analyzer, or other.

### PR DESCRIPTION
**Pull request by andreasgerstmayr**
Coverity Scan has limits on the frequency of build submissions (https://scan.coverity.com/faq#frequency), so I've configured weekly scans for now. If there are less than 4 daily commits to the master branch, we could also run a scan on every commit to the master branch.

Resolves #46

### CODE REVIEW

**Review by bjosv — APPROVED** [review]
Very nice!

File: .github/workflows/coverity.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    - name: Build with Coverity
+      run: cov-build --dir cov-int make
+      working-directory: build
+
+    - name: Submit the result to Coverity
+      run: |
+        tar czvf hiredis_cluster.tgz cov-int
+        curl \
+          --form token=${TOKEN} \
+          --form email=andreas@gerstmayr.me \
```
Comment (zuiderkwast) [inline .github/workflows/coverity.yml:43]: @andreasgerstmayr do we need to add your token to the Nordix repo? Or should we use mine or @bjosv's email here instead?

**Review by zuiderkwast — COMMENTED** [review]
Nice! It's a pity that this is only visible to us though. Better would be automatic feedback to contributors on PRs, but for that we can perhaps use other analyzers later.

File: .github/workflows/coverity.yml (hunk 1 above)
Comment (andreasgerstmayr) [reply .github/workflows/coverity.yml:43]: Yes, you need to add a GitHub secret named `COVERITY_TOKEN` to the repository. The token is visible here: https://scan.coverity.com/projects/hiredis-cluster?tab=project_settings

I'm not 100% sure what the mail is used for (I only know it doesn't work without), but yes, you can change it to one of you instead. Thanks!

**Comment by andreasgerstmayr** [comment]
> Nice! It's a pity that this is only visible to us though. Better would be automatic feedback to contributors on PRs, but for that we can perhaps use other analyzers later.

We can set the access to mode to "readonly for all" (Project Settings -> Edit -> Project Access). I'm not sure if that means public or for all logged in Coverity Scan users though.

Regarding PRs: Afaics there is no notion of multiple builds, I could only open the report of the last analyzed build. Coverity Scan looks quite old imho, but the static analysis of it is really great (and I don't know any free alternatives).

File: .github/workflows/coverity.yml (hunk 1 above)
Comment (zuiderkwast) [reply .github/workflows/coverity.yml:43]: * [x] COVERITY_TOKEN added

Should `--form token=${TOKEN}` be changed to `--form token=${COVERITY_TOKEN}`?

Perhaps each build is linked to a person. I don't mind it being you @andreasgerstmayr. I guess it has to be one of us listed here: https://scan.coverity.com/projects/hiredis-cluster?tab=members

Comment (andreasgerstmayr) [reply .github/workflows/coverity.yml]: > Should `--form token=${TOKEN}` be changed to `--form token=${COVERITY_TOKEN}`?

Done.

> Perhaps each build is linked to a person. I don't mind it being you @andreasgerstmayr. I guess it has to be one of us listed here: https://scan.coverity.com/projects/hiredis-cluster?tab=members

Yep, probably - I've updated it to your mail address now, I guess it should be one of the maintainers of this repo :smiley:

**Pull request merged**